Add DeviceIoControl

vbfox opened the pull request:
Multiple files were also move from the desktop project to the shared one
as file functions are present in both (CreateFile isn't but CreateFile2
is for example)

vbfox committed:
Add DeviceIoControl

Multiple files were also move from the desktop project to the shared one
as file functions are present in both (CreateFile isn't but CreateFile2
is for example)

vbfox commented on src/Kernel32.Shared/Kernel32+CreateFileFlags.cs:
```diff
@@ -146,13 +145,13 @@ public enum CreateFileFlags : uint
 
             /// <summary>Impersonates a client at the Anonymous impersonation level.</summary>
-            SECURITY_ANONYMOUS = SECURITY_IMPERSONATION_LEVEL.SecurityAnonymous << 16,
+            SECURITY_ANONYMOUS = 0 << 16,
```
One of the consequences of moving the file: I can't use `SECURITY_IMPERSONATION_LEVEL` anymore (AFAIK it's only used on desktop)

arlm replied:
How much of a problem is that? This SECURITY_IMPERSONATION_LEVEL enumeration does not even have their values assigned to them, they have been assigned automatically by the runtime. Is it something to make the code look better or to add some semantic only? shouldn't we always explicitly assign values to the enums that are used on the unmanaged APIs?

vbfox replied:
We tend to not specify the values when they are sequential especially when the native code header already use this facility: 

``` c
typedef enum _SECURITY_IMPERSONATION_LEVEL {
    SecurityAnonymous,
    SecurityIdentification,
    SecurityImpersonation,
    SecurityDelegation
    } SECURITY_IMPERSONATION_LEVEL, * PSECURITY_IMPERSONATION_LEVEL;
```

As far as re-using `SECURITY_IMPERSONATION_LEVEL` it's what the original header file does so it's more readable to port as close as possible but not essential. I was just pointing it out as one of the discutable point of the PR (Should I have used the computed value of 0 instead of `0 << 16` here for example.

AArnott replied:
> Should I have used the computed value of 0 instead of 0 << 16 here for example.

Ya, I was wondering the same thing. I think `0 << 16` is sufficiently enough of a head scratcher to defeat any case for the argument in its favor. I think 0 here is better.

> We tend to not specify the values when they are sequential especially when the native code header 

I hadn't noticed that. In a PR I review I usually ask for the values to be explicitly added, if nothing else than to make it evident that it wasn't simply overlooked (which it often is!) when they are important to have there.

vbfox replied:
> I hadn't noticed that. In a PR I review I usually ask for the values to be explicitly added, if nothing else than to make it evident that it wasn't simply overlooked (which it often is!) when they are important to have there.

Should we commit to this style and document it in contributing ?

AArnott replied:
I don't have a strong opinion. I sometimes wonder whether C# or C++ start enum values at 0 or 1 by default, and whether the two match on that behavior. Sometimes that question causes me to explicitly specify the values just to make sure they match. But if they match naturally (and I suppose they probably do) then I suppose as long as folks don't mind me asking in PRs whether the values are assigned in native code and we're missing them here, I guess we don't need to require it when implicit assignment is enough.

arlm replied:
@AArnott that questions bugs me also. I tend to prefer the explicit value as it mostly will be optimized if it is the same as the default one but increases readability by stating that it was something in purpose instead of something that was forgotten or overlooked.

AArnott replied:
>  it mostly will be optimized if it is the same as the default one 

I daresay it is exactly equivalent in IL and runtime regardless whether you explicitly specify values that in C# are considered the default.

arlm replied:
I do agree and think so also.

By optimized I meant disregarded by the compiler because it is the default value. It would only have readability meaning. I might have expressed myself on bad terms, but I don't know any better technical term to express this behavior other than "optimization" (the compiler disregarding the default values when they are explicit set).

vbfox commented on src/Kernel32.Shared/Kernel32+CreationDisposition.cs:
```diff
@@ -12,7 +12,7 @@ public partial class Kernel32
         /// </summary>
         /// <remarks>
-        /// These are flags to pass to the <see cref="CreateFile(string, ACCESS_MASK, FileShare, SECURITY_ATTRIBUTES*, CreationDisposition, CreateFileFlags, SafeObjectHandle)"/> method's dwCreationDisposition parameter.
+        /// These are flags to pass to the CreateFile method's dwCreationDisposition parameter.
```
We can re-link it if we add `CreateFile2` as it's what API sets include.

AArnott replied:
I'm good either way.

AArnott commented on src/Kernel32.Shared/Kernel32+CreateFileFlags.cs:
```diff
@@ -146,13 +145,13 @@ public enum CreateFileFlags : uint
 
             /// <summary>Impersonates a client at the Delegation impersonation level.</summary>
-            SECURITY_DELEGATION = SECURITY_IMPERSONATION_LEVEL.SecurityDelegation << 16,
+            SECURITY_DELEGATION = 3 << 16,
```
This seems odd though. We're not willing to declare the constants at this project, but we're willing to set the bits anyway? If the bits supported on non-desktop platforms, why can't we define the enum? And if the bits aren't supported, why are we using them?

vbfox replied:
I think we have some decision to take here.

Header files are structured with `WINAPI_FAMILY_PARTITION` macros such that both the values and the original enum (`SECURITY_IMPERSONATION_LEVEL`) are accessible from store apps even if they can't actually occur (AFAIK, `SECURITY_IMPERSONATION_LEVEL` is documented desktop only).
I suspect that it's there for ease of porting code without having to ifdef every usage of the unsupported values.

Should we:
- Mirror the header files ? (And present to store apps values and potentially whole enums that can't be used)
- Use `#ifdef` to try to signal as much as possible what is or isn't available ?
- Do what I did there (maybe in a better way) importing values but not whole new types ?

AArnott replied:
I think we should mirror the header files. Not as a strict rule since omitting an enum when it can't be used in a store app seems reasonable. But certainly in cases like this where we're losing semantically rich information and hard-coding values as a result, I think the least bad option is to go ahead and define the constants/enums in Store if necessary so that other things can depend on them.

arlm replied:
I like the idea of the constants, it keeps the semantics and the value without the need to put comments or XML Docs to explain the hard-coded values.

vbfox committed:
Merge branch 'master' of github.com:AArnott/pinvoke into deviceiocontrol

vbfox committed:
Move SECURITY_IMPRESONATION_LEVEL to Shared proj

As discussed in PR #269 while this enumeration isn't usable for anything
outside of the Desktop profile it's actually available in Non-Desktop
header files and referenced by other enums (CreateFileFlags).

So we mirror the header files and include it in the shared project.

AArnott commented:
LGTM. Thanks.
